Add git hooks and simplify scripts

lipis opened the pull request:
- [x] Add git hooks to make sure the files are formatted before commiting
- [x] Simplify `format*` scripts
- [x] Drop prettier-eslint (is just a wrapper for ESLint and Prettier that we are using already)

----
### The actual formats will happen in follow up PRs

- [Format preview for `format:code`](https://github.com/gatsbyjs/gatsby/pull/8623/files)
- [Format preview for `format:other`](https://github.com/gatsbyjs/gatsby/pull/8622/files)

lipis commented on www/src/data/StarterShowcase/generatedGithubData/gatsby-advanced-blog.json:
```diff
@@ -1 +1,195 @@
-{"name":"gatsby-advanced-blog","version":"0.1.9","description":"Gatsby starter for advanced blog","main":"n/a","author":"wonism","license":"MIT","homepage":"https://github.com/wonism/gatsby-advanced-blog","keywords":["gatsby"],"repository":{"type":"git","url":"git+https://github.com/wonism/gatsby-advanced-blog.git"},"bugs":{"url":"https://github.com/wonism/gatsby-advanced-blo/issues"},"peerDepend ...
```
we might want to ignore those.. or just format them as well if it's not an issue.

pieh commented:
I think we would need to consolidate husky things - we do have it setup for `www` https://github.com/gatsbyjs/gatsby/pull/8229 right now, (but at least for me - this setup in `www` works globally for repo - so not sure if it's my issue or just how it works) - we probably would want `husky` setup in single place - repo main package.json

DSchau commented:
@pieh I like the idea of using something like Husky and I _completely_ agree that it should be at the root of the repo. It seems like a bug or something that the one at www/ level is applying to _all_ files.

I'd rather be explicit and define exactly what files we auto format.

jquense commented on package.json:
```diff
@@ -45,27 +46,33 @@
-    "format-yaml": "prettier --write \"**/*.y?(a)ml\"",
+    "format": "npm-run-all -p format-packages format:js format:other",
+    "format:other": "yarn prettier --write",
+    "format:js": "eslint --ignore-path .gitignore \"**/*.{js,jsx}\" --fix",
```
why specify the the ignore path here?

lipis replied:
Because otherwise it will try to parse more files as far as I know if it's not included..

lipis replied:
also let's put more ignored directories (in another file) instead of specifying which ones to format... there is no need to have 10 different format rules, as it's the same project.

lipis replied:
I'm open for better names here

jquense commented on package.json:
```diff
@@ -45,27 +46,33 @@
     "plop": "plop",
     "prebootstrap": "yarn",
+    "precommit": "lint-staged",
+    "prettier": "prettier \"**/*.{json,md,scss,yaml,yml}\"",
```
I more descriptive name for this script would be good I think.

lipis replied:
any ideas? it's just a prettier command with all the supported files (except js, because it's handled with ESLint/Prettier)

jlengstorf replied:
`format-code` is nice and descriptive. How about that?

lipis replied:
But it's an actual `prettier` command.. and it formats everything except the actual code :)

lipis replied:
lint and format are doing different things, and the difference is the `--write` and `--list-different`, so if we remove that we will have to duplicate the supported files and it's better to have that in as few places as possible.

DSchau commented:
@lipis also you can ignore the commit hooks with [`git commit --no-verify`](https://stackoverflow.com/a/7230886) if that helps!

lipis commented:
I'll do that..

lipis commented on .eslintrc.json:
```diff
@@ -31,26 +31,12 @@
-        "exports": "always-multiline",
-        "functions": "ignore"
-      }
-    ],
```
Remove all style related rules from ESLint as Prettier is taking care of them

pieh replied:
will commit hooks run when someone is making changes via github edit UI / github merge conflicts resolver view? if not it would be beneficial to actually keep those I think

lipis replied:
We will have the CI tests for us if the linting is failing

lipis commented on .prettierrc:
```diff
@@ -1,5 +1,6 @@
-    "tabWidth": 2
+    "singleQuote": true,
+    "tabWidth": 2,
+    "trailingComma": "es5"
```
sort by name

lipis commented on package.json:
```diff
@@ -16,23 +16,24 @@
     "npm-run-all": "4.1.2",
     "plop": "^1.8.1",
-    "prettier": "^1.12",
-    "prettier-eslint-cli": "4.7.1",
```
no need to for prettier-eslint, we can simply use eslint with the config/plugin installed.. more flexibility.

lipis commented on www/package.json:
```diff
@@ -94,14 +94,6 @@
-  "husky": {
-    "hooks": {
-      "pre-commit": "pretty-quick --staged"
-    }
```
one source of truth for the hooks in the main package.json

lipis commented:
I updated the PR with only the changes and simplified package.json.

As a first step we can run the `format:other` to format all the files except js

DSchau commented on package.json:
```diff
@@ -45,27 +46,33 @@
-    "format-integration": "prettier-eslint --write \"integration-tests/**/src/**/*.js\" \"integration-tests/**/cypress/**/*.js\"",
-    "format-yaml": "prettier --write \"**/*.y?(a)ml\"",
+    "format": "npm-run-all -p format-packages format:js format:other",
+    "format:other": "yarn prettier --write",
```
Can we do `npm run prettier --write` here? I'd like to minimize the amount of places we require yarn explicitly!

lipis replied:
I would say we could get rid of npm-run-all and use yarn everywhere since we have `yarn.lock` and no `package-lock.json`

DSchau replied:
I disagree a lot with that! We want to keep barrier to entry low, so a change like that just makes that barrier higher and harder to get to for new contributors.

lipis replied:
That's invisible to the users and we already using yarn in the 'prebootstrap' and defined as engine..

DSchau replied:
I can imagine a user running `npm run lint:other` and being confused that fails. In that scenario, we made it harder for that user to contribute, and I'm not sure if it was for a real benefit.

lipis replied:
Updated it

lipis commented:
Here is the PR for formatting all files except JS (just as an example)

lipis commented on .eslintrc.json:
```diff
@@ -31,26 +31,12 @@
     "no-inner-declarations": "off",
     "valid-jsdoc": "off",
     "require-jsdoc": "off",
-    "quotes": ["error", "backtick"],
```
Prettier is taking care of the quotes.. and do we really want to use back-ticks for all quotes?!

NickyMeuleman replied:
Prettier doesn't convert backticks to quotes. https://github.com/prettier/prettier/issues/2996#issuecomment-335137564
Backticks are used across the entire codebase even when there is nothing to interpolate, also in the documentation. This was a deliberate decision: see https://github.com/prettier/prettier/issues/54 for discussion.

pieh replied:
documentation part I can agree on - we shouldn't "enforce" back-ticks there, but this shouldn't mean we need to change code style in non-docs part of the repo, right?

pieh replied:
I know stuff like this can get people polarized, but can we not "sneak in" change like that with this PR? Can we just setup hooks to format things here and have separate PR or issue to discuss code style changes?

lipis replied:
Kept that option.. just sorted the rules :)

NickyMeuleman commented:
It looks like you plan to use `eslint-plugin-prettier` to integrate Prettier into the ESLint config (to make ESLint format `.js` files with Prettier) and are using `eslint-config-prettier` to prevent issues where the 2 tools want to do something else.

This is great! 👍 
I wrote about [how to wire that up](https://nickymeuleman.netlify.com/blog/automagically-lint/)

lipis commented:
I also spent a fair amount of time with Prettier and ESLint configurations.. Here is my take on that: https://github.com/lipis/prettier-setup

lipis commented on package.json:
```diff
@@ -16,23 +16,24 @@
-    "prettier-eslint-cli": "4.7.1",
+    "prettier": "^1.14.3",
     "rimraf": "^2.6.1",
     "yargs": "^10.0.3"
```
Should I remove maybe the `^` from everywhere, while we are on it?

NickyMeuleman commented:
Should we report what `eslint-config-prettier` picks up to ESLint as an error/warning by adding `"prettier/prettier": "error"` to the `"rules"` in `eslintrc`? 🤔

This is a very dividing discusssion.
It means users that have an ESLint extension installed for their editor will see underlines for things like empty whitespace etc.

lipis commented:
We should so there is always only one way to write things.. I'll update later..

DSchau commented on package.json:
```diff
@@ -1,40 +1,41 @@
-    "@babel/preset-flow": "^7.0.0",
-    "@babel/preset-react": "^7.0.0",
-    "@babel/runtime": "^7.0.0",
+    "@babel/core": "7.0.0",
```
Is this solving something you were running into, i.e. an npm vs. yarn issue?

lipis replied:
No not really.. but pinned versions have less surprises..

lipis replied:
fair enough..

lipis replied:
@KyleAMathews I reverted them

lipis commented:
Anything else that should be part of this PR?

Obviously the test will fail and we will do the actual linting/formatting in the following PRs..

pieh commented:
> Anything else that should be part of this PR?
> 
> Obviously the test will fail and we will do the actual linting/formatting in the following PRs..

You still sneak in your personal preference for code style here, so that's a blocker right now

lipis commented:
@pieh Which one..?

pieh commented:
quotes (single) + not using backtics outside of imports

lipis commented:
I didn't notice the double quotes, sorry! and as for the backticks I simply removed all the stylistic options from eslint as per suggestion from Prettier when combined..

pieh commented:
> and as for the backticks I simply removed all the stylistic options from eslint as per suggestion from Prettier when combined..

Right, but this does change formatting settings for our code if someone doesn't use backticks - so this would be valid (and inconsistent):

```js
const a = `foo`
const b = "bar"
```

We need eslint quotes rule anyway for that stuff to be consistent. I don't feel strongly about formatting itself, but we should limit the scope of PR to just add formatting commit hooks, because sneaking change to code style (this PR doesn't mention that) is just not good idea.

lipis commented:
Re-added the quotes.. it can be discussed in a separate PR :)

lipis commented:
Even without it, that setting is not updating the quotes anyway: https://github.com/lipis/gatsby/pull/2/files

pieh commented:
> Even without it, that setting is not updating the quotes anyway

That's why we need quote rule in eslint anyway - if we decide to change code style (from backticks to quotes) in any future PR, eslint quote rule would need to enforce that. That's why leaving code style entirely to prettier sounds really good, but doesn't really work in practice

lipis commented:
So I think everything is ready, for this PR.. The main goal is to simplify the format scripts and lint/format all supported files.

pieh commented:
I was testing this locally and couple of thoughts:
- for some reason husky git hooks weren't installed for me (needed to do some manual hook install) - I suspect this has something to do with previous hooks installed by husky (one in `www/package.json`). Also previous husky was `^1.0.0-rc.14` (stable 1.0.0 was released 2 days ago - might be good idea to use that?)
- we need to expand eslint ignore list - I needed to add at least those entries:
  ```
  # eslint will ignore by default just root node_modules
  **/node_modules/**

  # right now this just ignore files for built gatsby package - need better rule to ignore built files for other packages
  packages/**/dist

  # ignore public dirs for websites
  www/public
  examples/**/public
  benchmarks/**/public
  integration-tests/**/public

  # that's not our code, we don't want to lint it
  flow-typed
  ```
  otherwise `format:js` would format a lot of stuff we don't want it to format - there are more things we would want to ignore - like some test fixtures (but this can be done on package level) and built files that are placed in package dirs
- I think it would be good idea to document this - probably https://github.com/gatsbyjs/gatsby#a-note-on-how-this-repository-is-organized and https://github.com/gatsbyjs/gatsby/blob/master/docs/docs/how-to-contribute.md#contributing-to-the-repo
- pre-commit hooks will reject commits that don't pass linting (which makes perfect sense), but it does enforce some git workflows we didn't had before - we really only care that final proposed PR pass linting (because we squash and merge PRs anyway) - any thoughts on that? (people can still use `--no-verify` flag during commits to workaround this if this gets in the way more than it helps)

lipis commented:
yes there is a new release and I need to update husky accordingly.. 

on mobile right now though :)

NickyMeuleman commented:
> pre-commit hooks will reject commits that don't pass linting (which makes perfect sense), but it does enforce some git workflows we didn't had before - we really only care that final proposed PR pass linting (because we squash and merge PRs anyway) - any thoughts on that? (people can still use --no-verify flag during commits to workaround this if this gets in the way more than it helps)

The linting/formatting could be done (via the npm scripts) during CI?
Would adding another test or integrating it into an existing test like the circleCI/travis/appveyor test work?

This would still allow users to commit PRs like they would before, but reject ones that don't pass the linting.

pieh commented:
> The linting/formatting could be done (via the npm scripts) during CI?

We do have CI lint test - see `ci/circleci: lint` in merge box, but not format

I should probably expand a little why I wanted to bring this up: it's not uncommon to open WIP PRs to get feedback on some early work. Rejecting commits that don't pass lint (stuff that `--fix` won't fix - such as unused variables) is kind of getting in the way. Workaround with `--no-verify` disables pre-commit formatting altogether.  Essentially I want to have cake and eat it too :D I don't think there is any good solution for that, but it doesn't hurt to ask

lipis commented:
@pieh I updated Husky and added more to the ignore list..

> pre-commit hooks will reject commits that don't pass linting (which makes perfect sense), but it does enforce some git workflows we didn't had before - we really only care that final proposed PR pass linting (because we squash and merge PRs anyway) - any thoughts on that? (people can still use --no-verify flag during commits to workaround this if this gets in the way more than it helps)

I don't think that this will be an issue, quite many folks already using Prettier/ESLint/Husky in other projects, and in the end of the day if there is an error that means that something couldn't be fixed and needs a human intervention anyway :) Of course `--no-verify` is always an option as well

> Could the linting/formatting be done (via the npm scripts) during CI?

Yes, but we don't want the CI to make commits.. the CI will point to the errors and it will be quite obvious what needs to be fixed :)

oorestisime commented:
FWIW I am not entirely sold on pre-commit hooks for a few reasons

* extra barrier for first contributions (as if people can contribute and in the end it will take a few minutes to fix eslint issues when merging the PR)
* it happens way to early in the process (as @pieh said in for wip PR)
* strictly personal opinion, i don't like tools autoformating my code :D I prefer having a tools to say what is wrong and adjust accordingly.

lipis commented:
> extra barrier for first contributions (as if people can contribute and in the end it will take a few minutes to fix eslint issues when merging the PR)

I don't see it as a barrier, especially the first contributors will have to comply with our coding style, and even if they are going to do a contribution via the GitHub UI, our CIs will spot the error if any. It will be up to us to fix (run the formatter) before merging. We have to approve/test anyway :)

> it happens way to early in the process (as @pieh said in for wip PR)

It's not an issue, as usually people don't write messy code even in the beginning. We have to comply with ESLint rules at any given time. If we don't like something, we just have to change the ESLint rules.

> strictly personal opinion, i don't like tools autoformating my code :D I prefer having a tools to say what is wrong and adjust accordingly.
Merge state

Let's leave to the robots what robots can do best.. and let's leave humans to spend time on the actual problem.. not how to format the code!

jquense commented:
I'd add in that i've mostly turned this on for all our repos and it's been a net good. The `--no-verify` hook is a lot less onerous than other things (like commit signing) partly because the hook tells you how to skip the validation when it fails. e.g. "didn't pass add --no-verify" to skip". The nice thing here as well is that is perfectly fine to skip verification for commits during a WIP pr, since (unlike signing) you don't need to go back and rewrite the git history to make it pass later, just fix the errors in a new commit. IMO the its the git history rewrites that make something really off putting to new contributors, since it's hard and scary for those not super familiar with git. This however is a good bit simpler since a dev already knows how to commit code, adding an additional flag isn't so bad

lipis commented:
Couldn't agree more!

lipis commented:
Just a reminder on how it will look like: [`format:js`](https://github.com/lipis/gatsby/pull/2/files) and [`format:other`](https://github.com/lipis/gatsby/pull/1/files).

We'll need to through the `✖ 369 problems (369 errors, 0 warnings)` for JS.

octalmage commented on package.json:
```diff
@@ -45,27 +46,37 @@
+  },
+  "husky": {
+    "hooks": {
+      "pre-commit": "lint-staged"
```
So this will prevent a user from committing things to their own branch? Is there a way to prevent husky from installing the hook? To me it seems like you should be able to commit whatever you want while working on something, then of course before merging and reviewing you have CI to catch this kind of stuff.

lipis replied:
In practise there is no harm..

jlengstorf replied:
@octalmage this only catches style problems and prevents a lot of "fix style" commits by flagging things ahead of time. In most cases, style fixes can be fixed automatically, so it shouldn't add much friction.

Of course, you can also always use `--no-verify` when committing to bypass this altogether (which the check will tell you if something fails).

NickyMeuleman commented:
I really like git-hooks as a tool to enforce standards. I also agree with points made above against them. 🤔 

Doing linting/formatting checks during CI seems like a good middle road.
It allows PRs that don't pass the checks. Which is fine, think about them like PRs with failing unit-tests. A PR with failing tests can exist, but doesn't get merged into master until they pass.

As @pieh suggested, whichever option is chosen, it has to be documented.
Instructions to format your PR could live in the PR-template. And/or a bot could comment on the PR if it doesn't pass the check in CI.

The React repo implemented this last year:
https://github.com/reactjs/reactjs.org/issues/67

An interesting option I read about to _enforce_ the linting/formatting standards while requiring zero effort is the git `pre-recieve` hook.
https://git-scm.com/book/pl/v2/Customizing-Git-Git-Hooks#_code_pre_receive_code
This is a serverside hook. That means `husky` will not support it.
This is over my head right now, not sure if it provides a solution for this problem. Or if GitHub even supports that specific hook.

All of this being said, I still like the `pre-commit` hook via `husky`. The `--no-verify` flag doesn't seem like it poses a huge hurdle for contributors.
(maybe move to `pre-push` since that is less frequent and hooks do take some time to run, this will help with PRs that have a lot of commits)

lipis commented:
> Doing linting/formatting checks during CI seems like a good middle road.
It allows PRs that don't pass the checks. Which is fine, think about them like PRs with failing unit-tests. A PR with failing tests can exist, but doesn't get merged into master until they pass.

But this is what will happen anyway.. if they get the message locally (something that it shouldn't) if they still want to commit pass the `--no-verify`.. The PR won't pass for the same reasons and since we have to fix that before merging to master.. why not asking (automatically) the users to fix those?!

If we see that there are many complains we can reconsider later, but if we document it, there is nothing hard.. just RTFM.

LeKoArts commented:
Today and yesterday the pre-commit hook on `www` really annoyed me. I'm working on a tutorial for the docs in my fork on a separate branch but it still messes with my commits. I commit often while being in the WIP status and didn't know until seeing this PR that this hook was the source of the headaches. Yes, the code formatting is not alright at the moment but I'm working on it!

Documenting the `--no-verify` flag won't do it. I use Webstorm or GitKraken to commit, other users use different software and it all functions a bit differently then.

I don't want my code to be messed with until at least I opened up a PR against the original branch. If the PR is open and I commit to my fork I'm fine with it (I like the CI idea) but not in my WIP fork.

lipis commented:
yes but this is a broken code and I think we just have to document on how to install prettie and ESLint in the editors..

LeKoArts commented:
I don’t understand your answer, sorry. I’m using eslint and prettier but this line is from a code block in a markdown file and it needs to be exactly formatted like I intended it to.
And as I said you then need to explain the commit flag for every major editor and I need to remember checking that checkbox for every commit.

We have often first time contributors here who e.g. only want to contribute to the docs (maybe mostly via the Github UI). And when they then commit such a code line like I did then will have no clue why that automatically happened. Maybe the tag after the three backticks was wrong but they won’t know.

lipis commented:
Your API_KEY example should look like this, because even though it's an example it should be a valid code, and in your case it wasn't!

```js
API_KEY='your-secret-access-token-wou7evoh0eexuf6chooz2'
```

You can configure editors to format all the supported formats that Prettier is supporting (js, md, scss, yaml, less, etc) and doing that on save will save a bunch of stuff.

Overall, using git hooks it's not that scary as many people think in this thread, they are used in quite a few popular projects with many new comers. It's all safe.

lipis commented:
@KyleAMathews @pieh so what do you think about this PR?

shannonbux commented:
@pieh and @lipis just thinking this discussion is long enough, it ought to be turned into an RFC! https://github.com/gatsbyjs/rfcs

pieh commented:
@shannonbux @lipis We discussed this today during maintainers meeting and concluded we don't want commits to be rejected on failed linting - it would more annoying than helpful. Formatting in git hooks would be fine, so I will be researching that part (I guess that's matter of using `prettier-eslint` instead of `eslint --fix`)

lipis commented:
prettier-eslint is exactly the same thing and can report errors if it can't fix the things that is trying to fix..

I'm still confused on why people are afraid of it, or just give it a try and see it in real life, how good/bad it is.. My guess is that the code quality will be much better overall and if there is a rejection during commit, the message is pretty clear.

jlengstorf commented:
It may be worth not running Prettier against Markdown files. A .env file regularly looks like this:

```sh
MY_API_KEY=ab-cd-12350980
```

That's valid in `.env`, and shouldn't be touched by Prettier. If it touches it in `*.md` files, that's a problem.

Any non-zero number of false positives will make this more cumbersome than useful, so I'd vote for scaling it back to avoid those possibilities.

LeKoArts commented:
@jlengstorf
It worth be noting that it only transformed my env file when I used e.g. the javascript highlighting for the code block. If I leave that away it won’t touch the code block.
But if I hadn’t remember that I wrote it like that on purpose I would never have catched that automatic change — and that’s why I say that these falsw positives will sure distract (first time) contributors

lipis commented:
again I don't see that as a problem because it was marked as js in markdown and in the end of the day that's why we have humans to review the final code. we spend a lot of time at Prettier to eliminate the false positives and not break the code.

jquense commented:
I think it may be help to break down pros/cons with the commit hooks. On one hand, as noted, its more of an "opt out" approach by the contributor to avoid having to fix lint issues during a WIP pr. The main benefit tho is that it's run _automatically_ whereas without them a user would need to remember to  run lint/prettier/format after they are done with their PR and consider it in working order.

In my experience the latter rarely happens, and leads to manual fixing of editing and lint issues after waiting on the really long CI loop to see what was wrong. On the other hand precommit hooks do require a contributor opt out of verification earlier in the process if they want to commit multiple times, but it also does a much better job of: 

1. ensuring finished PR's don't languish b/c a contributor didn't phase out at the end before fixing lint/formatting
2. Help the contributor get more immediate feedback on whether the PR will fail CI, saving them plenty of time if they don't remember to run these checks themselves when done.

I'm pretty savy with OSS flows and i never remember to run these tasks manually :P everyone of my React pr's is me committing, waiting 10min for CI and realizing I forgot a semicolon. In my opinion it's an equally off-putting experience.

DSchau commented:
@jquense I agree with a lot of what you're saying, but have a few reservations.

> In my experience the latter [running tasks manually] rarely happens, and leads to manual fixing of editing and lint issues after waiting on the really long CI loop to see what was wrong.

I don't think it's "really long" anymore :) The lint status reports back in 2-3 minutes, which seems like a pretty good way to get that feedback. That said, it can be more of a struggle for someone to _know_ that something failed and go check it out, rather than just presuming the PR is good, so I empathize with that. A CircleCI comment/webhook would be something worth investigating.

> Help the contributor get more immediate feedback on whether the PR will fail CI, saving them plenty of time if they don't remember to run these checks themselves when done.

I agree that they get immediate feedback. But that immediate feedback _could_ also be negative/confusing and it might cause he/she to just give up or grow frustrated. Whereas if we keep the barrier to PR low we can offer guidance as to how to get things formatted/fixed once it gets to the PR step.

> I'm pretty savy with OSS flows and i never remember to run these tasks manually :P everyone of my React pr's is me committing, waiting 10min for CI and realizing I forgot a semicolon. In my opinion it's an equally off-putting experience.

Agreed. But if we can reduce that time to say... 2-3 minutes, I think that's a good balance here.

---

> Currently if there are PRs with linting errors, it's not big deal for me and other maintainers to just fix up those problems and push to PR branch

@pieh I totally agree with this. It's one minute of our time vs. let's say 30 minutes or so of annoyance, increased barrier to PR  (i.e. someone may grow frustrated and not even submit the PR), etc.

---

TL;DR I like the value this PR provides to some users, but I'm worried about it causing issues for _other_ users. I won't fight too much either way though 🙃

jlengstorf commented:
Suggestion: what if we moved this to [`pre-push`](https://git-scm.com/docs/githooks#_pre_push) instead?

This solves both problems, as far as I can tell:

1. It doesn't interfere with the local development process at all.
2. It provides early feedback _before_ waiting for CI to complete.

Maybe this is a good way to get the benefits while avoiding the hurdles?

DSchau commented:
@jlengstorf that seems like a good compromise!

jquense commented:
I all with @pieh here i think, if we want a middle ground, running the check but not blocking is the best option.

That should be fine to do with lint-staged/husky, may need to swallow an exit code. We should at least run prettier on commit

I don't know that `pre-push` helps much for the WIP case right? often folks are pushing unfinished PR's for feedback or collboration

DSchau commented:
> We should at least run prettier on commit

That's kinda where @pieh and I ended up. I think we're fully in favor of that, definitely.

NickyMeuleman commented:
Could we run linting/formatting on `pre-commit` (or `pre-push`), but only thow a warning and not an error? (as @jquense suggested, changing an exit 1 somehow).
But do fail in CI.

This would allow the immediate feedback, while not preventing a commit.
The CI also prevents the final PR from getting merged with linting/formatting errors.

lipis commented:
Our first time contributors are still developers, they are not musicians, so dealing with linting errors is part of our job, especially when contributing to another project. I think I already said enough about it, but for once more, it's not so big of a deal, to fix the errors (that we have to deal with them anyway) at any stage of the PR, and it's much faster to get these errors earlier than wait 10 minutes for the CI.

LeKoArts commented:
> Our first time contributors are still developers

No, they are not. Enough people mentioned at various places that Gatsby is their first (bigger) JS/React project and that they learn both things in the process — in addition to that you forget people who contribute to the docs.

lipis commented:
> Enough people mentioned at various places that Gatsby is their first (bigger) JS/React project and that they learn both things in the process 

But they are developers and they will have to learn one more thing, there is nothing to be scared of when reading an error message and complying.. Obviously we will fix all the errors once this is merged. In the long run we will save so much time and have less errors in our CIs.

> in addition to that you forget people who contribute to the docs.
I'm guessing that if people updating the docs they are doing so from GitHub UI.. so nothing changes there.

pieh commented:
Just leaving update here - I've spent some time on this today and bugged some people - we will do test drive with git hooks. It proved impossible to just show errors without rejecting (`lint-staged` will swallow output if process exits with 0 code)

Will be looking to merge this over the weekend (to give us some time to merge as many PRs as possible to before merging this in to minimize conflicts caused by formatting)

We will add some additional custom message when commits will be rejected - something like this:
```
Gatsby uses `precommit` hooks to run our linting and style checks. We do this
to avoid additional hassle during pull request reviews, because all PRs must
pass these checks.

If you're doing something temporary, you can disable this hook with:
  git commit --no-verify

For more details, see gatsby.app/precommit-hooks
```
and on the linked page (doesn't exist yet) we will show instructions on how to opt out of git hooks if some people will find it really annoying.

As for actual PR - I spent some time on adjusting/updating ignored files - right now it looks like this for me:
```
*.min.js
**/node_modules/**
flow-typed

# built sites
benchmarks/**/public
e2e-tests/**/public
examples/**/public
integration-tests/**/public
www/public

# ignore built packages
packages/*/*.js
packages/gatsby/cache-dir/commonjs
packages/gatsby/dist

# fixtures
**/__testfixtures__/**
**/__tests__/fixtures/**
```

and consolidating eslintrc files (so we don't need one for each example site, etc)

lipis commented:
I updated the ignore list and updated the other two PRs to reflect what is being changed..

NickyMeuleman commented:
> It proved impossible to just show errors without rejecting
Tried this in https://github.com/gatsbyjs/gatsby/pull/8427#issuecomment-427397231 where it seemed to have worked (see last .gif). Did you get different results @pieh?

pieh commented:
@NickyMeuleman Ugh, my bad, you are right, I was just trying to use 0 exitcode in `lint-staged` rules instead of `husky` rules (joys of trying it late in the night :P). Will do some more checking later on, right now will focus on getting as much of other PRs in.

lipis commented:
@pieh  Why not prettify the JSON files as well.. and add to the ignore list if they are generated automatically?

pieh commented:
I missed weekend deadline as I needed to consult with some people working on `www` PRs (if resolving conflicts after running massive autoformat wouldn't cause them too much trouble).

So few things I just pushed here:
 - updated ignore rules to cover more things we don't want lint/format
 - removed .json linting/formatting (at least for now) - see commit msg for reasoning https://github.com/gatsbyjs/gatsby/pull/8427/commits/031ec68548b3dd6fd86b452ca60c4df5e38cbe97
 - updated lint-staged to version 8 (it adds support to lint/format partially staged files, so any changes that you don't actually commit don't need to pass linting)
 - added `hooks:(un)install` scripts to help uninstall and re-install hooks
 - and added not rejection (thanks @NickyMeuleman for help :) ) - this is how it looks like (message was changed a bit from when I took the screenshot):
![screen shot 2018-11-06 at 17 59 24](https://user-images.githubusercontent.com/419821/48085504-bdf7b000-e1fa-11e8-8f2f-08913acc4239.png)

lipis commented:
I also updated the #8622 and #8623

lipis commented:
> I didn't want to spend/waste more time on trying to figure it out - would prefer to merge this and if possible to fix that problem in follow up PR

Makes sense.. we can do it in the follow up!

pieh commented:
> I also updated the #8622 and #8623

Did you run "just" `format` on them or did you also fix manually any errors that can't be autofixed? I have branch https://github.com/gatsbyjs/gatsby/compare/master...pieh:hooks-plus-format where I did autoformatting (both code and other) + manual fixes to get linting pass with those changes

lipis commented:
I just run the code without any manual work.. it's just to see what is being affected and what is needed to be fixed.

I'll fix the errors and do manual updates when we merge the hooks.

pieh commented:
> I just run the code without any manual work.. it's just to see what is being affected and what is needed to be fixed.
> 
> I'll fix the errors and do manual updates when we merge the hooks.

Let's merge this PR, then 2 of your (if they will apply cleanly) and then we can merge remaining of manual fixes that I already did from my branch (no sense to add more work for You if it's already done)

lipis commented:
> Let's merge this PR, then 2 of your (if they will apply cleanly) and then we can merge remaining of manual fixes that I already did from my branch (no sense to add more work for You if it's already done)

Let's press the buttons! After merging this.. I will update my other PRs and force push the changes.. and then you can apply your changes..!!

Looking forward to pass all the tests!

pieh commented:
eta ~15 minutes for merge - just gave final "warning" for rest of team to get any changes that are complete merged (to avoid unneeded conflict resolutions for PRs that were ready but for some reason not merged).

pieh commented:
@lipis merged, can you update other PRs so they don't include same changes as this one (git is weird sometimes)

lipis commented:
Yep! Just did.. :)
